src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <credentials/GroupDataProvider.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/Pool.h>
+#include <platform/KeyValueStoreManager.h>
+#include <stdlib.h>
+#include <string.h>
+
+namespace chip {
+namespace Credentials {
+namespace {
+
+static constexpr size_t kIteratorsMax      = CHIP_CONFIG_MAX_GROUP_CONCURRENT_ITERATORS;
+static constexpr size_t kGroupNameMax      = CHIP_CONFIG_MAX_GROUP_NAME_LENGTH;
+static constexpr size_t kStoreKeyLengthMax = 128;
+static const char * kStorePrefix           = "gdp";
+
+struct FabricEndpoint
+{
+    uint32_t group_count;
+};
+
+class GroupDataProviderImpl : public GroupDataProvider
+{
+protected:
+    template <typename T, int S>
+    class Storage
+    {
+    public:
+        virtual ~Storage() = default;
+
+        CHIP_ERROR Save(const T & value)
+        {
+            T copy;
+            ReturnErrorOnFailure(DeviceLayer::PersistedStorage::KeyValueStoreMgr().Put<T>(this->mKey, value));
+            return DeviceLayer::PersistedStorage::KeyValueStoreMgr().Get<T>(this->mKey, &copy);
```

Comment:
Good observation, and I have put a comment here. The reason of this apparently superfluous read is a bug in the KeyValueStoreMgr implementation for Darwing. For some reason, if the key is new, the Get fails unless the Put is immediately follow by this otherwise-superfluous read. [I have opened a ticket to solve this problem](https://github.com/project-chip/connectedhomeip/issues/12174).